src/Compilers/CSharp/Test/Semantic/FlowAnalysis/FlowTests.cs

```diff
@@ -2034,6 +2034,694 @@ static bool Set(out int x)
+            CreateCompilation(source).VerifyDiagnostics(
+                // (8,15): error CS0165: Use of unassigned local variable 'x'
+                //             ? x.ToString() // 1
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "x").WithArguments("x").WithLocation(8, 15),
+                // (9,15): error CS0165: Use of unassigned local variable 'y'
+                //             : y.ToString(); // 2
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "y").WithArguments("y").WithLocation(9, 15),
+                // (24,15): error CS0165: Use of unassigned local variable 'x'
+                //             ? x.ToString() // 3
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "x").WithArguments("x").WithLocation(24, 15));
+        }
+
+        [Fact]
+        public void NullCoalescing_CondAccess_Throw()
+        {
+            var source = @"
+class C
+{
+    void M1(C c1, bool b)
+    {
+        int x;
+        _ = c1?.M(x = 0) ?? throw new System.Exception();
+        x.ToString();
+    }
+
+    bool M(object obj) { return true; }
+}
+";
+            CreateCompilation(source).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void NullCoalescing_CondAccess_Cast()
+        {
+            var source = @"
+class C
+{
+    void M1(C c1, bool b)
+    {
+        int x;
+        _ = (object)c1?.M(x = 0) ?? throw new System.Exception();
+        x.ToString();
+    }
+
+    C M(object obj) { return this; }
+}
+";
+            CreateCompilation(source).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void NullCoalescing_CondAccess_ImplicitReferenceConv()
+        {
+            var source = @"
+class C
+{
+    void M1(C c1, bool b)
+    {
+        int x;
+        object obj = c1?.M(x = 0) ?? (object)(x = 0);
+        x.ToString();
+    }
+
+    C M(object obj) { return this; }
+}
+";
+            CreateCompilation(source).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void NullCoalescing_CondAccess_BoxingConv()
+        {
+            var source = @"
+class C
+{
+    void M1(C c1)
+    {
+        int x;
+        object obj = c1?.M(x = 0) ?? c1.M(x = 0);
+        x.ToString();
+    }
+
+    bool? M(object obj) { return null; }
+}
+";
+            CreateCompilation(source).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void NullCoalescing_CondAccess_UserDefinedConv_01()
+        {
+            var source = @"
+struct S { }
+
+struct C
+{
+    public static implicit operator S(C? c)
+    {
+        return default(S);
+    }
+
+    void M1(C? c1)
+    {
+        int x;
+        S s = c1?.M2(x = 0) ?? c1.Value.M3(x = 0);
+        x.ToString(); // 1
+    }
+
+    C M2(object obj) { return this; }
+    S M3(object obj) { return this; }
+}
+";
+            // PROTOTYPE(improved-definite-assignment):
+            // Arguably, in this scenario, the RHS of the `??` should be visited in an unreachable state
```

Comment:
I'm not sure why this would be true, and I'm not sure why we error here. The spec says:

`Otherwise, if b has a type B and an implicit conversion exists from a to B, the result type is B. At run-time, a is first evaluated. If a is not null, a is unwrapped to type A0 (if A exists and is nullable) and converted to type B, and this becomes the result. Otherwise, b is evaluated and becomes the result.`

For this case, that means that we evaluate `c1?.M2(x = 0)`. If the result of that is not null, it is converted to `S`, and we know that `M2` had to be called, and `x` should be definitely assigned. If it was `null`, then `c1.Value.M3(x = 0)` is evaluated, which should definitely assign the result.

Reply:
Could you please link the relevant spec section so I can get more context?

Reply:
https://github.com/dotnet/csharplang/blob/main/spec/expressions.md#the-null-coalescing-operator

Reply:
After reading the spec, it feels like the implicit conversion on the left doesn't affect this analysis, since the null test always occurs before the left conversion. We just have to demonstrate that in the tests.